### PR TITLE
Fix several erros w.r.t. mixed_precision & edge case

### PR DESCRIPTION
**Pull request by LinjianMa**

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -2623,8 +2627,6 @@ def _cast_forward_inputs(self, *args, **kwargs):
         Moves the :meth:`forward` inputs to the compute device and casts them
         to the appropriate dtype if needed.
         """
-        if not self._is_root:
```
Comment (LinjianMa) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2626]: This is because otherwise `_cast_forward_inputs ` will not be called in the non recursive policy.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -2706,10 +2709,17 @@ def _register_post_forward_hooks(self):
         self._post_forward_handles.clear()
         for module in self.module.modules():
             module_param_handles = self._module_to_handles[module]
+            # reshard following handles: 1. the handle is in
+            # ``module_param_handles``and 2. the handle doesn't contain
+            # parameters in the parent module of the current module.
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py]: That is a great catch 😮

File: torch/distributed/fsdp/flat_param.py [hunk 3]
```diff
@@ -526,3 +526,13 @@ def _get_modules(self) -> Set[nn.Module]:
...
             set(spi.module for spi in self.flat_param._shared_param_infos)
         )
+
+    def _is_root(self, module: nn.Module) -> bool:
+        """
+        check if the input ``module`` is one of the root modules in
+        ``self._get_modules()``.
+        """
+        for m in self._get_modules():
+            if m is not module and module in list(m.modules()):
```
Comment (awgu) [inline torch/distributed/fsdp/flat_param.py]: Shall we factor out `list(m.modules())` into a `set`?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -1479,7 +1484,7 @@ def cast_fn(x: torch.Tensor) -> torch.Tensor:
             )
 
     @torch.no_grad()
-    def _cast_param_shards_to_dtype(self):
+    def _cast_param_shards_to_dtype(self, p: FlatParameter):
         """
         Allocates a mixed precision paramter shard and casts parameter shards to
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1489]: To-do: update the docstring to reflect that it now operates on a single `FlatParameter`.

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 5]
```diff
@@ -53,17 +54,40 @@
     "param_level": HandleInitMode.PARAM_LEVEL,
 }
 
+default_mp = MixedPrecision(
+    param_dtype=torch.float16,
+    buffer_dtype=torch.float16,
+    reduce_dtype=torch.float16,
+)
+
+# Params and buffers are not cast, comm only happens
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py:63]: It looks like this section is copied from `test_fsdp_mixed_precision.py`. We should de-duplicate it by refactoring it to `common_fsdp.py`. I can take care of this if you want.

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 6]
```diff
@@ -53,17 +54,40 @@
...
+mp_no_mixed_precision = MixedPrecision()
+
 
 class DoubleConv(nn.Module):
     def __init__(self) -> None:
         super().__init__()
         self.conv1 = nn.Conv2d(3, 6, 5)
         self.pool = nn.MaxPool2d(2, 2)
         self.conv2 = nn.Conv2d(6, 16, 5)
+        # Handles the edge chase when this parameter is wrapped together with
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]:
```suggestion
        # Handles the edge case when this parameter is wrapped together with
```

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 7]
```diff
@@ -271,12 +294,15 @@ def test_train_with_full_state_dict(
     @skip_if_lt_x_gpu(2)
     @parametrize("iters", [1, 3])
     def test_fsdp_flatten_params_exec_order(self, iters: int):
+        """
+        test the execution order of flatten parameters.
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]: Maybe we should make this a bit more specific. Is this testing the reverse gradient ready order path? If so, then maybe something like:
```suggestion
        Tests that ``FlatParameter`` s are reconstructed following reverse gradient
        ready order after the first iteration.
```
(rewrap accordingly)

Also, is this
https://github.com/awgu/pytorch/blob/2cdeefe0212fefe34ddcbe97b38e0afdcb66a531/test/distributed/fsdp/test_fsdp_param_exec_order_policy.py#L333
checking element-wise equality of the tensors in the list?

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 8]
```diff
@@ -271,12 +294,15 @@ def test_train_with_full_state_dict(
     @skip_if_lt_x_gpu(2)
     @parametrize("iters", [1, 3])
     def test_fsdp_flatten_params_exec_order(self, iters: int):
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]: If my understanding of what this test is testing is correct, then maybe:
```suggestion
    def test_reconstruct_reverse_gradient_ready_order(self, iters: int):
```
I am not too sure about that particular name and do not feel strongly about it.

However, the main point is that the suffix `params_exec_order` in the test name might become confusing since `torch.fx`, reverse gradient ready order, and possibly fake tensor are all approaches for getting the parameter execution order, while this test specifically exercises the reverse gradient ready order path.

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 9]
```diff
@@ -316,6 +342,42 @@ def test_fsdp_flatten_params_exec_order(self, iters: int):
         )
         self.assertTrue(fsdp_model._param_exec_order_state, ParamExecOrderState.INITIALIZED)
 
+    @skip_if_lt_x_gpu(2)
+    @parametrize("mp_config", [
+        default_mp, mp_only_reduce, mp_only_param_and_buf, mp_no_mixed_precision
+    ])
+    def test_fsdp_mixed_precision(self, mp_config: MixedPrecision):
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py:350]: Curious, did you write this test from scratch?

I am trying to think strategically about how to organize unit tests for the non-recursive wrapping path. It makes sense to have some coverage in `test_fsdp_param_exec_order_policy.py`, but where do the draw the line? (We still will need to add the non-recursive wrapping setting to the existing unit tests. I do not think we will test _every_ feature in this file.)

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 10]
```diff
@@ -316,6 +342,42 @@ def test_fsdp_flatten_params_exec_order(self, iters: int):
...
+        """
+        test the wrapping with mixed precisions.
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]: Sorry to nitpick, but again, I think we can make this more descriptive.
```suggestion
        Tests that ``FlatParameter`` dtypes are expected when using mixed precision.
```
(Again, I am not sure if we want to mention that this path is the reverse gradient ready order path. Maybe we can do that later when we integrate with `torch.fx`.)

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2626]: IIUC: The issue is that `_lazy_init()` is not called until `_pre_forward()`, which will only be run on the first module (that is not the `FullyShardedDataParallel` module). Then, `_cast_forward_inputs()` returns immediately since `self._is_root` is `None`.

Your change is quite clean 👍🏼

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 11]
```diff
@@ -2706,10 +2709,17 @@ def _register_post_forward_hooks(self):
         self._post_forward_handles.clear()
         for module in self.module.modules():
             module_param_handles = self._module_to_handles[module]
+            # reshard following handles: 1. the handle is in
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py]: nit:
```suggestion
            # Reshard the handles satisfying:
            # 1. the handle is in `module_param_handles` and
            # 2. the handle does not contain parameters in the parent module of `module`
```
(fyi: I single backtick for non-docstring comments since it saves space and does not render. I wanted to specify `module` instead of "the current module" to be clear that the current module refers to the exact variable `module` in the for loop.)

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 12]
```diff
@@ -2706,10 +2709,17 @@ def _register_post_forward_hooks(self):
...
+            # ``module_param_handles``and 2. the handle doesn't contain
+            # parameters in the parent module of the current module.
+            reshard_handles = []
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py]: nit: Do you think this is easier to read?
```suggestion
            handles_to_reshard = []
```

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 13]
```diff
@@ -3560,10 +3570,13 @@ def _rebuild_full_params(
                     and not force_full_precision
                 )
                 if mixed_precision_cast_ran:
-                    self._cast_param_shards_to_dtype()
-                    # TODO: remove below
-                    for p in self.params:
-                        assert p.dtype == self.mixed_precision.param_dtype
+                    self._cast_param_shards_to_dtype(p)
+                    p_assert(
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3574]: nice!

**Review by awgu — COMMENTED** [review]
Thanks for all the debugging! I really appreciate it.

As always, I left some minor comments. We can probably merge after you take a look.

File: torch/distributed/fsdp/flat_param.py (hunk 3 above)
Comment (LinjianMa) [reply torch/distributed/fsdp/flat_param.py]: done, let me know if this is good for you

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py (hunk 7 above)
Comment (LinjianMa) [reply test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]: yes

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py (hunk 8 above)
Comment (LinjianMa) [reply test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]: done

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py (hunk 10 above)
Comment (LinjianMa) [reply test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]: yes so this test doesn't involve reverse gradient ready order, so I don't think we need to mention that here
